### PR TITLE
Add official prettier settings 104

### PR DESCRIPTION
**Pull request by MattyMc**
## Linked Issue

#104 

## Description

- Updated prettier from `2.2.1` to `2.4.1`. Note, used the --exact flag as recommended by official prettier documentation, so you will no longer see the `^` in `package.json`.
- Added `.prettierrc` at root, set default options. 
- Ensured default setting from @calendee's [comment](https://github.com/Virtual-Coffee/virtualcoffee.io/issues/104#issuecomment-776498054) are captured in `.prettierrc`. 
- Installed `lint-staged` by following official prettier docs. Note, this has a dependency on `husky`, and the installation makes a few notable changes. In particular, adding a `.husky` config folder and a "lint-staged" property in `package.json.
- Prettier now runs on commit.
- Added a prettier ignore file (`.prettierignore`) and populated with contents of `.gitignore`, as recommended by prettier docs. 

Note, most changes to files are a result of the move from spaces to tabs. 

## Methodology

Followed documentation, experimented by running prettier from command line. Two helpful commands to run: 

 - `npx prettier --check .` - checks for any violations.
 - `npx prettier --write .` - detects and auto-fixes violations.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,65 +1,71 @@
...
+	"name": "virtualcoffee-io",
+	"version": "2.0.0",
+	"description": "The public site for virtualcoffee.io",
+	"scripts": {
+		"preinstall": "npx only-allow yarn",
+		"start": "npx grunt start",
+		"build": "ELEVENTY_ENV=production npx grunt release-prep",
+		"serve": "npx grunt connect:server",
+		"build-preview": "yarn build && yarn serve",
+		"prepare": "husky install"
```
Comment (danieltott) [inline package.json:11]: Let's add a `format` script here that runs prettier. Something like `"format": "npx prettier --write ."`

Comment (danieltott) [reply package.json:11]: Have it match the commands given to the `lint-staged` command

**Review by danieltott — CHANGES_REQUESTED** [review]
@MattyMc this is looking really good! I have one addition so far - I'll be able to take a closer look later on tonight. But This addition will allow users to run prettier via a yarn command without having to look up prettier's command line syntax 👍

File: package.json (hunk 1 above)
Comment (MattyMc) [reply package.json:11]: Great idea. I'll get this done shortly.

**Comment by MattyMc** [comment]
@danieltott I've added a "format" script to `package.json` as suggested. I'm only partially sure it's showing up with this PR... this is my first time editing a PR, and really only my third time making a PR to a public repo (i.e. newbie). Please let me know if you're not seeing this new commit.

I've added the "format" script as suggested, and ensured the same defaults are being passed as you suggested.

A few quick notes on what files are processed by prettier and when: 
- All files: Automatically when a commit is attempted.
- Only staged files: On demand when `npx lint-staged` is run. 
- All files: On demand when `npm run format` is run. We're passing the same defaults that `lint-staged` uses (see package.json): 
```json
"lint-staged": {
    "**/*": "prettier --write --ignore-unknown"
}
```

**Review by danieltott — CHANGES_REQUESTED** [review]
@MattyMc Let's try to format `njk` files as `html` files: according to the docs, we'd add these lines to the config file:

```
"overrides": [
	{
		"files": "*.njk",
		"options": { "parser": "html" }
	}
]
```

We might have to update the command to make sure `njk` gets picked up. We'd also have to hit up the fixes in #404 here instead of a separate PR.

But in my initial testing, I think this'll work out pretty well, and I'd love to have those files handled...

**Comment by MattyMc** [comment]
@danieltott yup we should definitely have those `njk` files handled somehow. Let me try it out and see how it goes here!

**Comment by MattyMc** [comment]
@danielcobo I've had a bunch of trouble trying to get prettier to do this through the config file. Here's what I've learned so far.

TLDR: It (seems) not possible to configure prettier to format a file extension `xyz` with parser `html` through prettier's config file. 

Long form:
- I found [this source](https://github.com/prettier/prettier-vscode/issues/1405) from the `prettier-vscode` repo that indicates that the config snippet you mentioned would work, but I think that's only for the vscode extension (not prettier). 
- [This issue](https://github.com/prettier/prettier/issues/8924) seems to confirm that prettier can't pick up arbitrary file extensions. The [prettier plugin docs](https://prettier.io/docs/en/plugins.html) indicate: "For additional languages you’ll need to install a plugin."
- It is possible to pass arguments through the command line, by (for example) running `npx prettier --check src/uses.njk`. Ironically, the CLI command picks up the parser from the config, but it doesn't seem possible to add the extension to the config. 

I have an idea for extending `lint-staged` to run using a custom glob as an argument. Might be a long shot, or may end up having to run two separate commands. I'll take a swing at it in the morning.

**Comment by danieltott** [comment]
@MattyMc This should be possible - what we'll want to do is add that overrides block to the code I pasted above - [see here for the docs on that](https://prettier.io/docs/en/configuration.html#setting-the-parserdocsenoptionshtmlparser-option)

Then, the problem is we're setting the `--ignore-unknown` flag, which was great before, but now we want it to try to process things. So, instead of collecting all files and ignoring unknown file types, we want to collect specific file types and process all the ones we pull in.

So our `lint-staged` command will become something more like this:

```
"**/*.{js, css, ts, scss, njk, md, yml, svg, json}": "prettier --write"
```

We'll need to make sure we've added all file types we want.

Then we can update the `format` command to do the same.

Thoughts?

**Comment by MattyMc** [comment]
Good morning @danieltott . I think I figure it out! 

I was hoping to get prettier to work natively, so that someone more familiar with prettier but maybe less so with the repo would be able to run generic prettier commands from the CLI and get expected results. For example: `npx prettier --check`, which would pick up on config settings, and also check `njk` files. I don't think this is possible though, which is okay. We can rely on our npm script: `npm run format`

I was able to get `lint-staged` to work as expected and to pass in a custom glob to prettier (I learned all about globs!). So now we pass prettier a custom glob like: `prettier ... "(**/*|**/*.njk)"`. It's looking good! Should have a PR up soon!

**Comment by danieltott** [comment]
@MattyMc sounds good - keep in mind we're using `yarn` and not `npm` 👍

**Comment by MattyMc** [comment]
@danieltott will do! Hey just quickly, I noticed this in the .editorconfig: `indent_style = space`. I'll update this to `tab` for consistency and include in my PR. [The docs](https://editorconfig.org/) say this is valid.

**Comment by MattyMc** [comment]
@danieltott all done! 

Our repo now: 

1. Auto-checks all staged files on commit for formatting. This includes `njk` files plus all the default file types. It will not allow a commit unless it can either correct a formatting issue (e.x. indentation), and will warn/prevent a commit if it cannot (e.x. closing tags). 
2. Prettier ignores files specified in the new `prettierignore`. 
3. `yarn run format` scans and attempts to correct all files in the repo. 
4. All files in the repo are now updated and aligned with correct formatting.

FYI in case there's currently unexpected behaviour on the site, it may have been from these types of things: 
Screen Shot 2021-10-05 at 9 39 59 AM

Everything fixed! Prettier works great! G2G!

**Review by danieltott — CHANGES_REQUESTED** [review]
@MattyMc this is looking great!

I noticed that it wasn't picking up fixes to json files, which made me realize we were missing some things in the `format` command.

I was able to get it to pick up all the files we want by changing it to this:

```
"format": "npx prettier --write --ignore-unknown --no-error-on-unmatched-pattern \"**/*\" && npx prettier --write \"**/*.njk\""
```

Basically one command to do the normal files, and one command to do `njk` files. If you have any better ideas I'm all ears too!

**Comment by MattyMc** [comment]
@danieltott thanks for catching that prettier is not picking up json files. Sorry I missed that!

Let me explore this a bit. The glob `(**/*|**/*.njk)`should first match all files prettier works with (**/*), and then explicitly all njk files (**/*.njk), and I thought JSON would be picked up in the first part of that. If not, I think extending the second part to `**/*.{njk,json}` might work, and have the added benefit of using a single CLI command (so all errors would be surfaced together).

I'll get into this shortly! 

Was adding the `--no-error-on-unmatched-pattern` flag this connected to not picking up the json files?

**Comment by MattyMc** [comment]
@danieltott this is strange. When I run... 

- `npx prettier --write --ignore-unknown "**/*"` I get `json`, `yml`, `njk`, `js`, `md`, `scss` files in the output. 
- `npx prettier --write --ignore-unknown "(**/*.njk|**/*)"` I get only `njk` and `md` <-- how I have it now in `package.json`
- `npx prettier --write --ignore-unknown "(**/*|**/*.njk)"` I get `json`, `yml`, `njk`, `js`, `md`, `scss` <-- arguments reversed in glob.

Strange behaviour. I want to investigate this further. Maybe I only need to make `njk` explicit with `lint-staged`, but I'll definitely test with all the file types.

**Comment by danieltott** [comment]
@MattyMc That first option does what we need I think. Maybe when you specify a file type in the prettier config, it stops being "unknown" ?

**Comment by MattyMc** [comment]
> @MattyMc That first option does what we need I think. Maybe when you specify a file type in the prettier config, it stops being "unknown" ?

Totally possible. The first option is definitely the way to go. I think where I was tripped up was that running `npx prettier .` and `npx prettier "**/*" are not equivalent. The former ignore `njk` files, and the latter includes them. Lost a little time on that one 😄 .

I'm going to update this now and do some manual checking to ensure prettier is running on all files through our three main workflows: 
- `yarn format`
- `npx lint-staged`
- automatically on commit, using husky/lint-staged

Will have the PR updated shortly! Thanks again for catching this!

**Comment by danieltott** [comment]
Thanks @MattyMc - this is looking awesome 💪

**Comment by MattyMc** [comment]
Thanks @danieltott 😄

**Review by MattyMc — COMMENTED** [review]
Not sure if I'm supposed to comment here (newbie), but these files look good to me!

**Review by danieltott — APPROVED** [review]
this is awesome
